### Order to Run Scripts ###

1) se_alignment.pl

2) extract_total_reads.py

3) exonic_read_counts.R

4) HTseq_counts.py

### Dependencies (some optional) ###

*Alignment*

TopHat: https://ccb.jhu.edu/software/tophat/tutorial.shtml
GenomicAlignments: https://bioconductor.org/packages/release/bioc/html/GenomicAlignments.html

*Read Counting*

HTseq: http://www-huber.embl.de/HTSeq/doc/install.html#install

*Differential Expression*

edgeR: https://bioconductor.org/packages/release/bioc/html/edgeR.html

*Visualization*

gplots: https://cran.r-project.org/web/packages/gplots/index.html

heatmap.3: https://github.com/obigriffith/biostar-tutorials/blob/master/Heatmaps/heatmap.3.R

*Gene Set Enrichment*

goseq: http://bioconductor.org/packages/release/bioc/html/goseq.html
